src/util/amount.rs

```diff
@@ -0,0 +1,876 @@
+        (self.as_sat() as f64) * 10_f64.powi(denom.precision())
+    }
+
+    /// Express this [Amount] as a floating-point value in Bitcoin.
+    ///
+    /// Equivalent to `to_float_in(Denomination::Bitcoin)`.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_btc(&self) -> f64 {
+        self.to_float_in(Denomination::Bitcoin)
+    }
+
+    /// Convert this [Amount] in floating-point notation with a given 
+    /// denomination.
+    /// Can return [ParseAmountError::TooBig] or [ParseAmountError::TooPrecise].
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn from_float_in(value: f64, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        // This is inefficient, but the safest way to deal with this. The parsing logic is safe.
+        // Any performance-critical application should not be dealing with floats.
+        Amount::from_str_in(&value.to_string(), denom)
+    }
+
+    /// Format the value of this [Amount] in the given denomination.
+    ///
+    /// Does not include the denomination.
+    pub fn fmt_value_in(&self, f: &mut fmt::Write, denom: Denomination) -> fmt::Result {
+        if denom.precision() > 0 {
+            // add decimal point and zeroes
+            let width = denom.precision() as usize;
+            write!(f, "{}.{:0width$}", self.as_sat(), 0, width = width)?;
+        } else if denom.precision() < 0 {
+            // need to inject a comma in the number
+            let nb_decimals = denom.precision().abs() as usize;
+            let real = format!("{:0width$}", self.as_sat(), width = nb_decimals);
+            if real.len() == nb_decimals {
+                write!(f, "0.{}", &real[real.len() - nb_decimals..])?;
+            } else {
+                write!(
+                    f,
+                    "{}.{}",
+                    &real[0..(real.len() - nb_decimals)],
+                    &real[real.len() - nb_decimals..]
+                )?;
+            }
+        } else {
+            // denom.precision() == 0
+            write!(f, "{}", self.as_sat())?;
+        }
+        Ok(())
+    }
+
+    /// Get a string number of this [Amount] in the given denomination.
+    ///
+    /// Does not include the denomination.
+    pub fn to_string_in(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_value_in(&mut buf, denom).unwrap();
+        buf
+    }
+
+    /// Get a formatted string of this [Amount] in the given denomination,
+    /// suffixed with the abbreviation for the denomination.
+    pub fn to_string_with_denomination(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_value_in(&mut buf, denom).unwrap();
+        write!(buf, " {}", denom).unwrap();
+        buf
+    }
+
+    /// Checked addition.
+    /// Returns [None] if overflow occurred.
+    pub fn checked_add(self, rhs: Amount) -> Option<Amount> {
```

Comment:
The `checked_*` stuff is great!

Reply:
I thought it was sad there are no traits for it, only in the `num` crate. But yeah it's good. We've been using it already internally.

Reply:
We could have an optional dependency on `num-traits`, I'm using it for `rust-lightning-invoice`and think it's an acceptable dependency in many cases.

Reply:
Just out of curiosity: how would you do an optional trait impl with a method signature identical to a native method of the type? 
```rust
#[cfg(feature = "num-traits")]
impl CheckedAdd for Amount {
  fn checked_add(self, other: Amount) -> Amount {
    self.checked_add(other)
  }
}
```
This looks like an infinite loop..

Reply:
I think you can differentiate them by calling `Amount::checked_add(self, other)` or `CheckedAdd::checked_add(self, other)`.